### DISCUSSION TITLE
Why foreground service when implementing background location listener

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13643] [Q&A]
Compare to [this](https://titaniumsdk.com/api/titanium/android/service.html#background-location-service-example) and [this](https://github.com/m1ga/titanium_background_app) example I implemented a app. So after click start button the `locationService.js` is executed once and terminates.

If I implemented without the foreground notification I got only two records in my debug file one by clicking start and one by clicking stop button.

But why I have to implement this `foregroundNotify` service to keep the location event listener alive? And how to remove the Notification from status bar? I got an error if I try to stop the service or call `Ti.Android.currentService` because the service seems already terminated.

index.js

```
function onOpen(){
    Ti.Geolocation.requestLocationPermissions(Ti.Geolocation.AUTHORIZATION_ALWAYS, e => {
    if (e.success) {

    } else {
      alert('Could not obtain location permissions.');
    }
  });
}

Ti.Geolocation.accuracy = Ti.Geolocation.ACCURACY_HIGH;
let intent = Titanium.Android.createServiceIntent( { url: 'locationService.js' } );
let service = Titanium.Android.createService(intent);

let f = Ti.Filesystem.getFile(Ti.Filesystem.applicationDataDirectory, "myDebug.txt");

function recordGeoPoint(e){
  let date = new Date();
   let dateFormat = '' + ("0" + date.getHours()).slice(-2) + ':' + ("0" + date.getMinutes()).slice(-2) + ':' + ("0" + date.getSeconds()).slice(-2);
  f.append("[" + dateFormat + "]  running... \n");
  }

function myStart(){
  f.write("");
  Ti.Geolocation.addEventListener('location',  recordGeoPoint );
  service.start();
}

function onClickStart(e) {
  if (parseInt(Ti.Platform.version.split('.')[0]) >= 10){
    Ti.Android.requestPermissions([ 'android.permission.ACCESS_BACKGROUND_LOCATION' ], e => {
      if (e.success) {
        myStart();
      } else {
        alert('Could not obtain background location permissions.' + e.error + "\n" + Ti.Platform.version.split('.')[0]);
      }
    });
  } else {
    myStart();
  }
}

function onClickStop(e) {
   Ti.Geolocation.removeEventListener('location',  recordGeoPoint );
//service.stop();
}
```

locationService

```
let channel = null;
if (Ti.Platform.Android.API_LEVEL >= 26) {
	channel = Ti.Android.NotificationManager.createNotificationChannel({
		id: "channel_id",
		name: "Background Channel",
		importance: Ti.Android.IMPORTANCE_LOW
	});
}

Ti.Android.currentService.foregroundNotify(
	123,
	Ti.Android.createNotification({
		contentTitle: "Background app",
		contentText: "Keep the app running...",
		channelId: channel ? channel.id : null,
		contentIntent: Ti.Android.createPendingIntent({
			intent: Ti.App.Android.launchIntent || Ti.Android.createIntent(),
		})
	})
);

```

**Comment by m1ga** [comment]
The foreground service is needed because of Android. Check https://developer.android.com/guide/components/foreground-services for all the restrictions you have. I think it started in Android Oreo that you can't start a service in the background that runs longer than 10secs without a foreground notification.
In Android 13 you are able to swipe them away and they'll be in the bottom:
<img src="https://developer.android.com/static/images/guide/components/fgs-manager.svg"/>

The system can kill the app when it is in the background if needed. The foreground notification is there to keep it alive.

**Reply by beppo-ivel** [reply]
And how to cancel the notification if you run the service (locationService.js) only once? Or do you set the service interval to 1h? Does use a interval, which is executed every second more computational power then a interval which is executed every 1h or is it quite similar?
